external/vcm/vcm/limit.py

```diff
@@ -0,0 +1,113 @@
+            variables are limited
+
+    """
+
+    def __init__(
+        self, alpha: float, limit_only: Optional[Sequence[str]] = None,
+    ):
+        self._alpha: float = alpha
+        self._limit_only: Optional[Sequence[str]] = limit_only
+        self._limits: xr.Dataset = None
+
+    def fit(
+        self,
+        ds: xr.Dataset,
+        feature_dims: Optional[Sequence[str]] = None,
+        fit_indexers: Optional[Mapping[str, int]] = None,
+    ) -> "DatasetQuantileLimiter":
+        """Fit the limiter on a dataset.
+        
+        Args:
+            ds: Dataset to be used to fit the limits.
+            feature_dims: Dimensions along which quantile limits should NOT be
+                computed, i.e., the resulting limits will be computed separately
+                for each coordinate along the feature_dims.
+            fit_indexer: Integer-based indexers that select a subset of `ds` on
+                which to fit limits. Indexed dimensions must not be in `feature_dims`.
+            
+        Returns: Fitted limiter
+        
+        """
+        if fit_indexers is not None and feature_dims is not None:
+            for index_dim in fit_indexers.keys():
+                if index_dim in feature_dims:
+                    raise ValueError(
+                        f"Indexer dim {index_dim} may not be in feature_dims."
+                    )
+
+        sample_ds = (ds.isel(**fit_indexers) if fit_indexers is not None else ds).load()
+        sample_dims = (
+            set(sample_ds.dims) - set(feature_dims)
+            if feature_dims is not None
+            else sample_ds.dims
+        )
+        lower = sample_ds.quantile(self._alpha / 2.0, dim=sample_dims)
+        upper = sample_ds.quantile(1.0 - self._alpha / 2.0, dim=sample_dims)
+        self._limits = xr.concat(
+            [lower, upper],
+            dim=xr.DataArray(["lower", "upper"], dims=["bounds"], name="bounds"),
+        )
+        return self
+
+    def transform(
+        self, data: Union[xr.Dataset, xr.DataArray]
+    ) -> Union[xr.Dataset, xr.DataArray]:
+        """Limit data.
+        
+        Args:
+            data: Dataset or dataarray to be limited
+            
+        Returns: Limited dataset or dataarray
+        
+        """
+        if self._limits is None:
+            raise ValueError("Limiter method .fit must be called before .transform")
+        if isinstance(data, xr.Dataset):
+            limited = data.copy()
+            vars_ = self._limit_only if self._limit_only is not None else data.data_vars
+            for var in vars_:
+                limited[var] = _limit_extremes(data[var], self._limits[var])
+        elif isinstance(data, xr.DataArray):
+            if (self._limit_only is None) or (data.name in self._limit_only):
+                limited = _limit_extremes(data, self._limits[data.name])
+            else:
+                limited = data
+        return limited
```

Comment:
Does this API need to be polymorphic or is it just a small convenience for e.g. interactive use? In addition to making this function more complicated, returning union types can make life difficult since any downstream code won't know if `.transform` returned  a dataset or dataarray. You can type hint this more effectively using `TypeVar`, but I think you should just delete the DataArray support. This will fail anyway if `array.name` is not defined.

Users can manually cast the array to dataset like this if they want: `arr.to_dataset(name="the name")`.

```suggestion
    def transform(
        self, data: xr.Dataset) ->xr.Dataset:
        """Limit data.
        
        Args:
            data: Dataset to be limited
            
        Returns: Limited dataset
        
        """
        if self._limits is None:
            raise ValueError("Limiter method .fit must be called before .transform")
     
        limited = data.copy()
        vars_ = self._limit_only if self._limit_only is not None else data.data_vars
        for var in vars_:
           limited[var] = _limit_extremes(data[var], self._limits[var])
        return limited

Reply:
Switched to dataset input as suggested